### PR TITLE
Add support for defining Vault address and token through environment variables in Consul Connect CA provider

### LINKED ISSUE #16853: Provide VAULT_TOKEN env variable support.
#### Feature Description
Let consul use VAULT_TOKEN env variable instead of in config.

#### Use Case(s)
Security. So that we can easily change the mode / perms of the env variable instead of the whole config file.

### PR DESCRIPTION
**Pull request by thevilledev**
### Description

- Add support for defining the Vault token from the `CONSUL_CA_VAULT_TOKEN` environment variable for the Consul Connect CA provider. If the token is provided through the environment variable, it takes precedence over the value set in the configuration file.

- Add support for defining Vault address from the `CONSUL_CA_VAULT_ADDR` environment variable for the Consul Connect CA provider.  If the address is provided through the environment variable, it takes precedence over the value set in the configuration file.

This PR also adds validation and tests for the Vault address. Previously it was possible to start the provider with an empty address, leading to runtime errors such as:

```
[ERROR] connect.ca: Failed to initialize Connect CA: routine="CA initialization" error="error configuring provider: Get \"/v1/auth/token/lookup-self\": unsupported protocol scheme \"\""
```

The parameter is marked as a required one already in the provider documentation.

### Testing & Reproduction steps

- Follow the [Connect CA getting started guide](https://developer.hashicorp.com/consul/tutorials/vault-secure/vault-pki-consul-connect-ca)
- Set the Vault token through `CONSUL_CA_VAULT_TOKEN` environment variable for the Consul process.
- Comment out the `token` field from the Consul configuration file upon testing.
- Verify that Consul Connect CA provider is able to interact with Vault with the token set through the environment variable.

Same test for `CONSUL_CA_VAULT_ADDR` environment variable and the `address` config file parameter.

### Links

Fixes #16853. 

### PR Checklist

* [x] updated test coverage
* [x] external facing docs updated
* [x] appropriate backport labels added
* [x] not a security concern

### CODE REVIEW

File: website/content/docs/connect/ca/vault.mdx [hunk 1]
```diff
@@ -93,7 +93,7 @@ The key after the slash refers to the corresponding option name in the agent con
   This token must have [proper privileges](#vault-acl-policies) for the PKI
   paths configured. In Consul 1.8.5 and later, if the token has the [renewable](/vault/api-docs/auth/token#renewable)
   flag set, Consul will attempt to renew its lease periodically after half the
-  duration has expired.
+  duration has expired. You can also provide this token through an environment variable `VAULT_TOKEN`. Token has to be provided either through the configuration file or through the environment variable. Both cannot be used at the same time.
```
Comment (trujillo-adam) [inline website/content/docs/connect/ca/vault.mdx:96]:
```suggestion
  duration has expired. You can also provide the token through the `VAULT_TOKEN` environment variable. The token must be provided either through the configuration file or through the environment variable, but you cannot provide the token using both methods to at the same time.
```

**Review by trujillo-adam — COMMENTED** [review]
Just a few tweaks on the docs. Someone from engineering will have to review and approve the code.

**Comment by jkirschner-hashicorp** [comment]
@thevilledev : Is the PR description still accurate? I noticed that it mentions:
> Remove the token field from the Consul configuration file.

We don't want to remove the token field at this time (and its removal would be a breaking change). Luckily, if I'm reading your PR correctly, it seems like the code doesn't actually remove the token field. Maybe the description needs to be updated accordingly?

I'll discuss with engineering and ask them to review.

**Comment by jkirschner-hashicorp** [comment]
@thevilledev : What do you think about also adding support in this PR for pulling the Vault address from the `VAULT_ADDR` environment variable?

I realize that's probably a less meaningful improvement than support for `VAULT_TOKEN`, but it seems like another standard Vault environment variable that applies here. (I'm also not sure if vaultapi does something under-the-hood to pull in `VAULT_ADDR` if the [Address is not explicitly specified in vaultapi.Config](https://github.com/hashicorp/consul/blob/ba402bbfddc0b091eb5cb45ee8f46611c7426186/agent/connect/ca/provider_vault.go#L110-L112)).

File: website/content/docs/connect/ca/vault.mdx (hunk 1 above)
Comment (thevilledev) [reply website/content/docs/connect/ca/vault.mdx:96]: Thanks!

**Comment by thevilledev** [comment]
Hey @jkirschner-hashicorp! Thanks for the comments.

> We don't want to remove the token field at this time (and its removal would be a breaking change). Luckily, if I'm reading your PR correctly, it seems like the code doesn't actually remove the token field. Maybe the description needs to be updated accordingly?

Sorry, probably poor wording from my side! I only meant the token field should be removed from the active configuration file while configuring the token through env variables. I'll update the description just to make this clear!

> What do you think about also adding support in this PR for pulling the Vault address from the VAULT_ADDR environment variable?

Great idea, I'll add this too. It's a small change as `vaultapi.NewClient()` does read the env variables if `config.Address` is not set upon initialisation. Code path looks like this:

- [`vaultapi.NewClient()`](https://github.com/hashicorp/vault/blob/main/api/client.go#L602-L603)
- [`vaultapi.DefaultConfig()`](https://github.com/hashicorp/vault/blob/main/api/client.go#L249)
- [`config.ReadEnvironment()`](https://github.com/hashicorp/vault/blob/main/api/client.go#L270)

I can add this into the PR a bit later today :+1:

**Comment by thevilledev** [comment]
Both `VAULT_TOKEN` and `VAULT_ADDR` are now part of this PR. Updated the description as well.

I noticed that the address validation was missing previously. Launching the agent without an address gives the following:

```
[ERROR] connect.ca: Failed to initialize Connect CA: routine="CA initialization" error="error configuring provider: Get \"/v1/auth/token/lookup-self\": unsupported protocol scheme \"\""
```

With this PR it's mandatory to set it either in the config file or as an environment variable. Added tests for it as well. I believe this is the correct behaviour, as [the documentation](https://developer.hashicorp.com/consul/docs/connect/ca/vault#address) already states that the address is a required parameter.

But let me know what you folks think! I hope this doesn't blow up the scope of the PR too much :)

**Review by kisunji — COMMENTED** [review]
Looks good at first glance! I started the CI to run tests.

cc @jkirschner-hashicorp One edge case to consider: this change could potentially break setups of users who already had `VAULT_*` environment variables since those will now cause the CA to error on startup. They could simply remove the address from the config but what if `VAULT_ADDR` must be different from what the CA provider should point to?

**Comment by jkirschner-hashicorp** [comment]
> One edge case to consider: this change could potentially break setups of users who already had VAULT_* environment variables since those will now cause the CA to error on startup. They could simply remove the address from the config but what if VAULT_ADDR must be different from what the CA provider should point to?

@kisunji : Fair point. That means explicit non-empty config values should override their respective `VAULT_*` environment variables. However, if I'm reading the PR correctly, it seems like the code currently prioritizes the environment variables over the config values.

If we make an explicit token or auth method in the config always override the environment variable, then it shouldn't be a breaking change I think... If it was working before, it would still work (because the existing settings would override the environment variable).

Do you agree with the above?

**Comment by kisunji** [comment]
The general convention is to have environment variables as higher priority over config files so it may be confusing if we decide to make the config address override the environment variable. 

I think an alternative solution is to make new env vars `CONSUL_VAULT_CA_ADDR` `CONSUL_VAULT_CA_TOKEN` which shouldn't overlap with any existing variable names. This way we don't accidentally couple Consul's operations with Vault's well-known environment variable names.

**Comment by jkirschner-hashicorp** [comment]
@thevilledev : I'm curious about your perspective as a Vault + Consul user on the environment variable precedence / naming conversation above.

**Comment by thevilledev** [comment]
Good discussion!

In terms of precedence, I belive at least consul-template first reads the configuration from a file and only if a value is not set, it proceeds to read it from an environment variable (or sets an empty/non-nil value). I think Vault does the opposite. These are not that well documented, so from a user PoV you find these things by trial-and-error or by reading the code.

I would vote for what @kisunji proposed. Two new environment variables and treat them with higher priority than what's set in the config file. If any current or upcoming Consul feature then requires a Vault integration the same approach will work. Both implicitly relying on the same set of environment variables might not work for all cases. Vault address might be different, for example, so a single `VAULT_ADDR` environment variable would be problematic.

For naming convention I would still have the postfix as "VAULT_ADDR" and "VAULT_TOKEN". Maybe something like `CONSUL_CONNECT_CA_VAULT_ADDR` and `CONSUL_CONNECT_CA_VAULT_TOKEN`?

**Comment by kisunji** [comment]
Maybe we can shorten to `CONSUL_CA_VAULT_ADDR` and `CONSUL_CA_VAULT_TOKEN`? I'm in full support of this approach in any case!

**Comment by jkirschner-hashicorp** [comment]
`{prefix}_CA` has the advantage of allowing us to distinguish between the agent TLS CA and the service mesh (connect) CA in the future, were we to build a more direct integration between the Vault PKI secrets engine and the agent TLS CA system (e.g., present credentials to Vault, get back an agent TLS cert).

I'd prefer "MESH" as the prefix rather than "CONNECT". We try to use "mesh" instead of "connect" in public docs now, except where knowledge of / reference to the fact that the subsystem in the code that provides more mesh capabilities is called "connect" for historical reasons.

Ultimately, I think the following forms are all OK, and have mixed feelings about which is best:
- `CONSUL_CA_VAULT_TOKEN`
- `CONSUL_MESH_CA_VAULT_TOKEN`
- `CONSUL_CONNECT_CA_VAULT_TOKEN`

**Comment by thevilledev** [comment]
Thank you for the context @jkirschner-hashicorp! I implemented it now with `CONSUL_CA_VAULT_TOKEN` and `CONSUL_CA_VAULT_ADDR` based on your comments and updated the ssue description. Changed the logic also to have env variable precedence over config file + tests for it.

**Comment by jkirschner-hashicorp** [comment]
@thevilledev : I'll make sure this gets a review in the next week. It looks like this slipped through the cracks.
